bundles/core/src/main/java/com/adobe/cq/commerce/core/components/models/retriever/AbstractCategoryRetriever.java

```diff
@@ -260,12 +260,21 @@ public String generateQuery(String identifier) {
      * @return GraphQL query as string
      */
     public Pair<CategoryListArgumentsDefinition, CategoryTreeQueryDefinition> generateCategoryQueryArgs(String identifier) {
-        FilterEqualTypeInput identifierFilter = new FilterEqualTypeInput().setEq(identifier);
+        FilterEqualTypeInput identifierFilter = new FilterEqualTypeInput().setEq(identifier.replaceAll("_", "/"));
```

Comment:
Can any part of the identifier contain underscores? If so we have to consider escaping. 

Actually I would recommend to use dashes instead of underscores.

https://developers.google.com/search/docs/advanced/guidelines/url-structure?hl=en&visit_id=637568438766532832-2808360326&rd=1

> Consider using punctuation in your URLs. The URL http://www.example.com/green-dress.html is much more useful to us than http://www.example.com/greendress.html. We recommend that you use hyphens - instead of underscores _ in your URLs.

Reply:
The url format and with that at least the `url_key` and `url_path` are controlled by the commerce backend. For Magento https://docs.magento.com/user-guide/catalog/catalog-urls.html, other solutions might have different formats.

So while we technically can totally change this I think we still want to align with what is configured for the backend.

Reply:
You mean, for Magento parts of the identifier (=url_key) may contain dashes right? If so we cannot use dashes for Mangeto but as you said others may do that differently. May be the char to replace the path separator needs to be configureable? 

However do we distinguish between author and publish? For publish a suffix should work better than a selector and there we can have slashes. If not mixing underscores in the url is not a good idea because Google joins words on them while splitting on dash. SEO is irrelevant for author.

Reply:
>Maybe the char to replace the path separator needs to be configurable?
Yes good idea, however we want to make the entire URL handling simpler. So config with default is ok to me.

>For publish a suffix should work better than a selector and there we can have slashes.
We use selectors because of caching behavior on the AEM dispatcher. Plus there is typically some URL rewriting setup on the publish as well. The public visible URL will typically not use any selector or suffix.

Reply:
Suffixes can as well be cached on dispatcher:

/path/to/category-page.html => editing purpose, 400 Bad Request on publish
/path/to/category-page.cat.html/url_path.html

Selectors' main purpose is view selection (they are taken into account for script resolution) and not data transport. I agree that using suffix on author is tricky in the editor but they are much more suited to transfer information (esp. hierarchical information)

Handling the rewriting/SEO handling of urls on dispatcher level spreads logic across systems which makes the system harder to maintain as a whole. (Think about e2e tests that would require a dispatcher to exists).

Reply:
> Suffixes can as well be cached on dispatcher

I know but requires extra config if I'm right?

> Handling the rewriting/SEO handling of urls on dispatcher level spreads logic across systems which makes the system harder to maintain as a whole. (Think about e2e tests that would require a dispatcher to exists).

I totally agree here, however it is still a common practice to use /etc/map and mod_rewrite.

Reply:
I'll use ~ instead of _
Looks like Magento doesn't allow ~ in URL key definitions so we are safe

Reply:
According to RFC1738 Section 2.2 is a unsafe character and we must encode it. Let's use with the underscore for now.